docs/markdown/snippets/asm.md

```diff
@@ -0,0 +1,18 @@
+## New language `nasm`
+
+When the `nasm` language is added to the project, `.asm` and `.nasm` files are
```

Comment:
I really would prefer to have `.nasm` only for NASM, at least for the time being. It's always safer to be strict first and relax the requirements later.

If we want to add MASM support later, we would have the same issue where preexisting projects use `.asm` for MASM sources. And then introduce the special case that MASM conflicts with NASM, or having to add `suffixes:` to `add_language` or other horrible things.

Let's keep Meson opinionated. :)

Reply:
> I really would prefer to have `.nasm` only for NASM

That would make this PR basically useless, I never seen any project using that extension. They would have top copy all their .asm files to the builddir and rename them. I'm trying to follow what projects actually do, and `.asm` for NASM is by far the most common.

> If we want to add MASM support later, we would have the same issue where preexisting projects use `.asm` for MASM sources. And then introduce the special case that MASM conflicts with NASM, or having to add `suffixes:` to `add_language` or other horrible things.

I think that's a non-issue actually. MASM is itself much more rare than NASM, and I would be very surprised any project mix nasm and masm at the same time, for the same platform. We could make both of them compile `.asm` files and just make adding those languages mutually exclusive.

Reply:
I agree that it would be better to just use `.nasm` for NASM, we can always relax it later if really needed but the other way around not…

Reply:
> They would have top copy all their .asm files to the builddir and rename them.

Why would they need to copy their source files instead of just rename them once they adopt native meson nasm support?

IMHO ideally we would support `.asm` for which meson will pick either nasm or yasm, `.nasm` for which meson will only use nasm and `.yasm` for which meson will only use yasm, but I am not sure how hard it would be to fit that into how meson currently handles languages/extensions…

Reply:
@xclaesse I don't care about what projects do *now*, I only care if projects can adapt. It wouldn't be the first time Meson forces you to follow Meson conventions, and that's a feature, not a bug.

Reply:
> Why would they need to copy their source files instead of just rename them once they adopt native meson nasm support?

Yes but only for projects that actually adopt meson, not wrapdb ports for example. To be fair, projects I would like to use it are all non upstream meson ports: libjpeg-turbo, ffmpeg, libffi.

> IMHO ideally we would support `.asm` for which meson will pick either nasm or yasm, `.nasm` for which meson will only use nasm and `.yasm` for which meson will only use yasm

It's true that currently it would fallback to use yasm to compile .nasm file which could be surprising. I would like to keep yasm and nasm as the same language at the meson level, because it's pretty much like gcc/clang, 2 compilers for the same language even if not 100% compatible. To be honest, I'm getting tempted to actually remove `.nasm` completely, I don't want projects to start renaming their files to `.nasm` if they can actually be built with yasm too. If a project really supports only yasm or only nasm, they should rather do:
```
asm_id = meson.get_compiler('nasm').get_id()
if asm_id != 'yasm'
  error('YASM compiler is required but found' + asm_id)
endif
```
That's basically the same situation for a project that requires GCC and not CLANG, or vice versa.

That being said, assigning .nasm to only NASM and .yasm to only YASM is easy to do from an implementation POV. I'm just not sure there is an actual use-case for it...

Reply:
I think you overestimate how compatible yasm/nasm nowadays are, it's I do not think it is uncommon that projects work just with one or the other. While this is frequent for yasm/nasm it is rare for C compilers.

So either we should:

- Make `.nasm` and `.yasm` be limited to nasm/yasm respectively
- Drop `.nasm` and `.yasm` extensions altogether

Everything else will just be way too confusing…

Reply:
> I think you overestimate how compatible yasm/nasm nowadays are, I do not think it is uncommon that projects work just with one or the other. While this is frequent for yasm/nasm it is rare for C compilers.

gcc/clang was a bad example, they are mostly compatible. gcc/msvc is a better example.

Reply:
@ePirat @bonzini Updated the PR to only accept `.asm` suffix and build them with either nasm or yasm. We can easily add `.nasm` and `.yasm` suffixes later.

Reply:
fwiw, I agree that we shouldn't invent a new extension, and using nasm to handle `.asm` files is fine. If we want to change this later, we could add a new kwarg to `add_language()` which specifies which file extensions we want that compiler to handle or something like that. 

Picking a default doesn't mean it will be the only option going forward. Just means we can't change the default without a long deprecation process.